runtime/src/self_test.rs

```diff
@@ -0,0 +1,187 @@
+    run_self_test(verify_status).await;
+    // Dead code. ya_runtime_api::server::run_async requires killing a process to stop
+    Ok(())
+}
+
+pub(crate) fn verify_status(status: anyhow::Result<ProcessStatus>) -> anyhow::Result<String> {
+    let Ok(status) = status else {
+        bail!("Failed to get self test status: {err}");
+    };
+    if status.return_code == 0 {
+        return Ok(String::from_utf8(status.stdout)?);
+    }
+    match String::from_utf8(status.stderr) {
+        Ok(stderr) => anyhow::bail!(
+            "Process failed, code: {}, stderr: {stderr}",
+            status.return_code
+        ),
+        Err(err) => {
+            anyhow::bail!(
+                "Process failed, code: {}. Failed to parse err output: {err}",
+                status.return_code
+            )
+        }
+    }
+}
+
+pub(crate) async fn run_self_test<HANDLER>(handle_result: HANDLER)
+where
+    HANDLER: Fn(anyhow::Result<ProcessStatus>) -> anyhow::Result<String>,
+{
+    let work_dir = std::env::temp_dir();
+
+    let deployment = self_test_deployment(&work_dir)
+        .await
+        .expect("Prepares self test img deployment");
+
+    let runtime_data = RuntimeData {
+        deployment: Some(deployment),
+        ..Default::default()
+    };
+    let runtime = Runtime {
+        data: Arc::new(Mutex::new(runtime_data)),
+    };
+
+    server::run_async(|e| async {
+        let ctx = Context::try_new().expect("Creates runtime context");
+
+        log::info!("Starting runtime");
+        let (status_sender, mut status_receiver) = mpsc::channel();
+        let emitter = EventEmitter::spawn(ProcessOutputHandler {
+            handler: Box::new(e),
+            status_sender,
+        });
+        let start_response = crate::start(work_dir.clone(), runtime.data.clone(), emitter.clone())
+            .await
+            .expect("Starts runtime");
+        log::info!("Runtime start response {:?}", start_response);
+
+        let run_process: ya_runtime_sdk::RunProcess = server::RunProcess {
+            bin: "/ya-self-test".into(),
+            args: vec!["ya-self-test".into()],
+            work_dir: "/".into(),
+            ..Default::default()
+        };
+
+        log::info!("Runtime: {:?}", runtime.data);
+        log::info!("Self test process: {run_process:?}");
+
+        let pid: u64 = crate::run_command(runtime.data.clone(), run_process)
+            .await
+            .expect("Runs command");
+
+        let (final_status_sender, final_status_receiver) = tokio::sync::oneshot::channel();
+        tokio::spawn(async move {
+            let status = collect_process_status(&mut status_receiver, pid).await;
+            final_status_sender.send(status)
+        });
+        let process_result = final_status_receiver
+            .await
+            .expect("Receives process status");
+
+        log::info!("Process finished");
+        let result = handle_result(process_result).expect("Handles test result");
```

Comment:
`server::run_async` fails to stop so results produced by `handle_result` function are printed to stdout/err, and then process is killed with `std::process::exit(0)`. It is a shame it will look like my code, but it was already like this (and I suspect it was the reason we had separated `test` and `offer-template` commands - there was no nice way to print results from `test`, so `offer-template` has been created).